Implements vara-run command line tool

boehmseb opened the pull request:
This tool allows running benchbuild experiments directly from the tool suite.
It provides the (to us) relevant functionalities of benchbuild's `run`, `slurm`, and `container` commands in a unified interface and enables a simpler handling of paper configs. For example it checks whether selected projects and versions are contained in the current paper config and if no projects are specified, it runs the experiment on all case studies in the current paper config.

resolves se-sic/VaRA#777

boehmseb committed:
vara-run first draft

boehmseb committed:
Fix problems with relative paths and revision filtering

boehmseb committed:
Fix problems with creating fresh bb configs

boehmseb committed:
Fix a type error

boehmseb committed:
Finish vara-run implementation + tests

boehmseb committed:
Make vara-run project/version selection tests more meaningful

boehmseb commented:
We probably should also write a short tutorial/documentation before merging this.

boehmseb committed:
Make folder assignments during save to default values

boehmseb committed:
Fix a test

boehmseb committed:
Fix LOC calculation

boehmseb committed:
Merge branch 'f-FixRepoLOC' into f-vara-run

boehmseb committed:
Fix some pylint issues

boehmseb committed:
Remove unused code

boehmseb committed:
Docs and Tutorials

boehmseb committed:
Suppress some pylint warnings

boehmseb committed:
Fix missing argument in vara-run test

vulder committed:
Merge branch 'vara-dev' into f-vara-run

boehmseb committed:
Merge branch 'vara-dev' into f-vara-run

vulder committed:
Merge branch 'vara-dev' into f-vara-run

vulder commented on varats/varats/tools/driver_run.py:
```diff
@@ -0,0 +1,175 @@
+    "projects",
+    multiple=True,
+    callback=__validate_project_parameters,
+    help="Only run experiments for the given project."
```
```suggestion
    help="Only run experiments for the given project. "
```

vulder commented on docs/source/tutorials/getting_started.rst:
```diff
@@ -0,0 +1,57 @@
+After a successful install, we need to create a proper environment to run experiments in.
+First, we need to select a directory where we create our setup.
+From now on, we refer to this directory as the `VaRA-TS root`.
+**Unless noted otherwise, ``vara-*`` commands should always be run from this directory.**
```
```suggestion
**Unless noted otherwise,** ``vara-*`` **commands should always be run from this directory.**
```

vulder reviewed (changes requested):
- [x] Pass extra -vvvv flags
- [x] Maybe provide project list as last paramters (make projects arguments instead of options)
- [x] Stream output from BB to stdout
- [x] Add option for auto submitting (by default no submission, with option submission)

boehmseb committed:
Apply suggestions from code review

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

boehmseb committed:
vara-run: add verbosity flag

boehmseb committed:
vara-run: make project selection a variadic argument

boehmseb committed:
vara-run: forward benchbuild output to stdout

boehmseb committed:
vara-run: add flag for running sbatch

boehmseb committed:
Rename unused variables

vulder reviewed (approved):
Looks good to me. Let's wait for some feedback from our other testers and then merge it :grin:

Sinerum reviewed (approved):
Looks good to me. Test went smooth.

MMory commented:
The output of vara-run -vvvv is not consistent with benchbuild -vvvv - when compiling xz, vara-run does not print the output of the configure script and furthermore the coloring does not work.

boehmseb commented:
> The output of vara-run -vvvv is not consistent with benchbuild -vvvv - when compiling xz, vara-run does not print the output of the configure script and furthermore the coloring does not work.

As for the color, I think this is stripped because I'm using `tee` to forward the output. Not sure how I can circumvent this. The plumbum documentation doesn't say anything about that issue.
I have no idea however, why the output of configure is missing. Both, `stdout` and `stderr` should be forwarded.
I might have to ask the BenchBuild  dev what other tricks he uses there :thinking:

cormensratio reviewed (approved):
Test runs worked flawlessly. Lgtm

J03D03 reviewed (approved):
Tested the `JustCompile` experiment with the `vara-run` command line tool against several C projects.
Lgtm

boehmseb committed:
Implement proper handling of keyboard interrupts.

BenchBuild has a handler for this case and produces additional output.
To capture that output, we also need to install a handler for that case.
Unfortunately, this means reimplementing plumbum's TEE, since that gives
us no access to the underlying process.

vulder committed:
Merge branch 'vara-dev' into f-vara-run

boehmseb commented:
We have 2 smaller issues left:

se-sic/VaRA#797
se-sic/VaRA#798

boehmseb committed:
make mypy and pylint happy

vulder commented:
resolves se-sic/VaRA#779
